Quantum-Control-Applications/ESR/T1.py

```diff
@@ -0,0 +1,172 @@
+        I_st.buffer(len(waits)).average().save("I")
+        Q_st.buffer(len(waits)).average().save("Q")
+
+################################
+# Open quantum machine manager #
+################################
+
+qmm = QuantumMachinesManager(host=qop_ip, port=qop_port)
+
+
+#######################
+# Simulate or execute #
+#######################
+
+simulate = True
+
+if simulate:
+    # simulation properties
+    simulate_config = SimulationConfig(
+        duration=2000,
+        include_analog_waveforms=True,
+        simulation_interface=LoopbackInterface(([("con1", 3, "con1", 1), ("con1", 4, "con1", 2)]), latency=180),
+    )
+    job = qmm.simulate(config, T1, simulate_config)  # do simulation with qmm
+    job.get_simulated_samples().con1.plot()  # visualize played pulses
+
+    analog_wf = job.simulated_analog_waveforms()
+
+    ver_t1 = (
+        analog_wf["elements"]["ensemble"][2]["timestamp"] + (analog_wf["elements"]["ensemble"][2]["duration"] / 2)
+    ) - (analog_wf["elements"]["ensemble"][0]["timestamp"] + (analog_wf["elements"]["ensemble"][0]["duration"] / 2))
+    ver_t3 = (
+        analog_wf["elements"]["resonator"][0]["timestamp"] + (analog_wf["elements"]["resonator"][0]["duration"] / 2)
+    ) - (analog_wf["elements"]["ensemble"][2]["timestamp"] + (analog_wf["elements"]["ensemble"][2]["duration"] / 2))
+
+    print("center to center time between 1st and 2nd pulse", ver_t1)
+    print("center to center time between readout and 2nd pulse", ver_t3)
+
+else:
+    qm = qmm.open_qm(config)
+
+    job = qm.execute(T1)  # execute QUA program
+
+    res_handle = job.result_handles
+
+    I_handle = res_handle.get("I")
+    I_handle.wait_for_values(1)
+    Q_handle = res_handle.get("Q")
+    Q_handle.wait_for_values(1)
+    iteration_handle = res_handle.get("iteration")
+    iteration_handle.wait_for_values(1)
+    next_percent = 0.1  # First time print 10%
+
+    def on_close(event):
+        event.canvas.stop_event_loop()
+        job.halt()
+
+    f = plt.figure()
+    f.canvas.mpl_connect("close_event", on_close)
+    print("Progress =", end=" ")
+
+    while res_handle.is_processing():
+        try:
+            I = I_handle.fetch_all()
+            Q = Q_handle.fetch_all()
+            iteration = iteration_handle.fetch_all()
+            if iteration / n_avg > next_percent:
+                percent = 10 * round(iteration / n_avg * 10)  # Round to nearest 10%
+                print(f"{percent}%", end=" ")
+                next_percent = percent / 100 + 0.1  # Print every 10%
+
+            plt.plot(waits * 4, I)
+            plt.plot(waits * 4, Q)
+            plt.title(f"iteration: {iteration}")
+            plt.pause(0.2)
+            plt.clf()
+
+        except Exception as e:
+            pass
+
+    plt.cla()
+    I = I_handle.fetch_all()
+    Q = Q_handle.fetch_all()
+    iteration = iteration_handle.fetch_all()
+    print(f"{round(iteration/n_avg * 100)}%")
+
```

Comment:
Since you're also acquiring the raw traces, I would plot them to save some time next time we'll need it and comment the lines if we don't want these plots

Reply:
cool